### ISSUE TITLE
[<Ray component: Tune >] Unable to re-produce results of trials from Ray Tune

### ISSUE THREAD

**Issue by hungpthanh** [ray-project/ray#43900]
### What happened + What you expected to happen

**1. The bug**
Currently, I'm trying to reproduce the results of a trial from Raytune by using the same hyperparameter in the trial. However, when I re-train the model by using the same hyperparameter, the results cannot be reproduced. Besides, I also have set random seed in trainable function.

To be simple,
In my experience, I have fixed the param space (i.e., each hyperparameter has only one value) as below and the data is fake.
```
param_space = {
        "lr": 1e-4,
        "weight_decay": 1e-3,
}
```

And, I have utilised grid search from Raytune.
```
tuner = tune.Tuner(
        tune.with_parameters(train_tune),
        param_space=param_space,
        tune_config=tune.TuneConfig(
            max_concurrent_trials=1,
        )
    )

results = tuner.fit()
results.get_dataframe()
```
Therefore, in my setup, raytune will only evaluate only 1 trial.
In addition, The **_train_tune_** function will show loss and accuracy for each epoch (iteration).

**2. Expected behaviour**
I expect that the loss and accuracy obtained from the train_tune function when `tuner.fit()` is performed will be the same as that in the `train_tune` function when I just execute it without Ray Tune.

**3. Useful Information**
The log below is the log of loss and accuracy when I obtained it from raytune and executed `train_tune` function directly.
3.1. Log from Raytune
```
Epoch: 0, loss: 0.6192966848611832, acc: 0.43140358898159786
Epoch: 1, loss: 0.5539271265268326, acc: 0.4333261393265243
Epoch: 2, loss: 0.5361498117446899, acc: 0.4751926158861984
Epoch: 3, loss: 0.5245729327201843, acc: 0.4909079069464567
Epoch: 4, loss: 0.5164188235998154, acc: 0.4937039702891986
```
3.2. Log from execute `train_tune` function directly
```
Epoch: 0, loss: 0.6192966878414154, acc: 0.43140358898159786
Epoch: 1, loss: 0.5539270728826523, acc: 0.4333261393265243
Epoch: 2, loss: 0.5361498117446899, acc: 0.4751926158861984
Epoch: 3, loss: 0.5245729446411133, acc: 0.4909051443455611
Epoch: 4, loss: 0.5164188176393509, acc: 0.4937067415453582
```

When the number of epochs increases, the difference in the results can become larger.
Therefore, it is hard to reproduce the result from Raytune.

### Versions / Dependencies

Ray: 2.9.0
python: 3.8
PyTorch: 2.2.1
PyG: 2.4.0

### Reproduction script

```
import os.path as osp
import random

import numpy as np
import torch
import torch.nn.functional as F
import ray.train
from sklearn.metrics import f1_score
from torch_geometric.datasets import PPI
from torch_geometric.loader import DataLoader
from torch_geometric.nn import GATConv
from ray import tune

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
N_EPOCHS = 5


class Net(torch.nn.Module):
    def __init__(self, n_features: int, n_classes: int):
        super().__init__()

        self.conv1 = GATConv(n_features, 256, heads=4)
        self.lin1 = torch.nn.Linear(n_features, 4 * 256)
        self.conv2 = GATConv(4 * 256, 256, heads=4)
        self.lin2 = torch.nn.Linear(4 * 256, 4 * 256)
        self.conv3 = GATConv(4 * 256, n_classes, heads=6, concat=False)
        self.lin3 = torch.nn.Linear(4 * 256, n_classes)

    def forward(self, x, edge_index):
        x = F.elu(self.conv1(x, edge_index) + self.lin1(x))
        x = F.elu(self.conv2(x, edge_index) + self.lin2(x))
        x = self.conv3(x, edge_index) + self.lin3(x)
        return x


def train_tune(config):
    torch.manual_seed(0)
    np.random.seed(0)
    random.seed(0)
    path = osp.join(osp.dirname(osp.realpath(__file__)), "..", "data", "PPI")
    # global train_dataset, val_dataset
    train_dataset = PPI(path, split="train")
    val_dataset = PPI(path, split="val")

    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=2, shuffle=False)

    model = Net(train_dataset.num_features, train_dataset.num_classes).to(DEVICE)
    loss_op = torch.nn.BCEWithLogitsLoss()
    optimizer = torch.optim.Adam(
        model.parameters(), lr=config["lr"], weight_decay=config["weight_decay"]
    )

    for _ in range(N_EPOCHS):
        train_loss = train(model, optimizer, loss_op, train_loader)
        val_f1 = test(model, val_loader)
        ray.train.report({"loss": train_loss, "accuracy": val_f1})
        print("Epoch: {}, loss: {}, acc: {}".format(_, train_loss, val_f1))


def train(model, optimizer, loss_op, train_loader):
    model.train()

    total_loss = 0
    for data in train_loader:
        data = data.to(DEVICE)
        optimizer.zero_grad()
        loss = loss_op(model(data.x, data.edge_index), data.y)
        total_loss += loss.item() * data.num_graphs
        loss.backward()
        optimizer.step()

    return total_loss / len(train_loader.dataset)


@torch.no_grad()
def test(model, loader):
    model.eval()

    ys, preds = [], []
    for data in loader:
        ys.append(data.y)
        out = model(data.x.to(DEVICE), data.edge_index.to(DEVICE))
        preds.append((out > 0).float().cpu())

    y, pred = torch.cat(ys, dim=0).numpy(), torch.cat(preds, dim=0).numpy()
    return f1_score(y, pred, average="micro") if pred.sum() > 0 else 0


def main():
    param_space = {
        "lr": 1e-4,
        "weight_decay": 1e-3,
    }

    tuner = tune.Tuner(
        tune.with_parameters(train_tune),
        param_space=param_space,
        tune_config=tune.TuneConfig(
            max_concurrent_trials=1,
        )
    )

    results = tuner.fit()
    results.get_dataframe()

    train_tune(param_space)  # The results from raytune should be the same when executing this function directly


if __name__ == "__main__":
    main()

```

### Issue Severity

High: It blocks me from completing my task.

**Comment by hungpthanh** [comment]
Hi, I would appreciate any help with my case.
Thank you.

**Comment by woshiyyya** [comment]
Can you disable train dataloader shuffle and ensure fixed all random seeds (e.g. `pl.seed_everything()`)?
